### ISSUE TITLE
Question regarding Colour Scheme of Error Output

### ISSUE THREAD

**Issue by DEberhardt** [pester/Pester#1735]
## 1. General summary of the issue

_Drowning in a see of red, do I find the threads beginning?_

This is more a general question about colour schemes and may be daft, but is there a way to influence the Colour Scheme of Error-Messages? Passed Tests green and Failed tests Red is fine, however, the explanation below the failed text is the same colour as the first Line "[-] ..." and thus harder to find the described "It" to pinpoint which test has been violated.

**Example:**

<p><span style="color:red">[-] TeamsFunctions Module Tests.Testing Module PUBLIC Functions.'Get-TeamsOVP' should have a BEGIN, PROCESS and END block</span> 5ms (4ms|1ms)
 <span style="color:white">Expected 'begin {' to be found in file 'Get-TeamsOVP.ps1', but it was not found.
 at "$($_.FullName)" | Should -FileContentMatch 'begin {', TeamsFunctions.Tests.ps1:95
 at <ScriptBlock>, TeamsFunctions.Tests.ps1:95</span>

<span style="color:red">[-] TeamsFunctions Module Tests.Testing Module PUBLIC Functions.'Get-TeamsOVP' should be an advanced function</span> 5ms (4ms|1ms)
<span style="color:white">Expected 'cmdletbinding' to be found in file 'Get-TeamsOVP.ps1', but it was not found.
 at "$($_.FullName)" | Should -FileContentMatch 'cmdletbinding', TeamsFunctions.Tests.ps1:102
 at <ScriptBlock>, TeamsFunctions.Tests.ps1:102</span>
<span style="color:red">

[-] TeamsFunctions Module Tests.Testing Module PUBLIC Functions.'Get-TeamsTDP' should have a BEGIN, PROCESS and END block</span> 5ms (4ms|1ms)
 <span style="color:white">Expected 'begin {' to be found in file 'Get-TeamsTDP.ps1', but it was not found.
 at "$($_.FullName)" | Should -FileContentMatch 'begin {', TeamsFunctions.Tests.ps1:95
 at <ScriptBlock>, TeamsFunctions.Tests.ps1:95</span></p>

I tried to colour these with <span> Blocks, but the preview does not show it, so I am unsure whether I can _show_ the meaning...
I think what I am saying is that the "Expected ..." line is secondary information which would not _need to_ be red or could be any other colour (like dark red/grey/white)?

I did not find any settings I could change, so I thought I raise it here.

Cheers,
David

**Comment by nohwnd** [comment]
There are no settings, but you could add them (internally, or on the config object) and make a PR. I  would start from just hacking the change in (output.ps1) to see what you mean and posting few screenshots so we can talk about the change.

**Comment by asears** [comment]
@DEberhardt do you have any screen shots?
 
As part of the build scans, could look for hard-coded colour names or other hard-coded values which could be configured.  Not sure if that could be done with PSScriptAnalyzer rule or using Find cmdlet....

I cleaned up a few hard-coded colors for consistency and will push this PR.  $ReportTheme.Fail is the colour that seems to be used for failure messages.  It would need to be split up?  Also is the foreground supposed to be red?  

There's another library which might have some useful patterns.
https://github.com/EvotecIT/PSWriteColor

As the terminal colors may behave differently across Powershell Core & Windows, might be good to expose a pesterconfiguration for $ReportTheme as a config object, and provide ability to make them b&w (or dark themed pester) for accessibility and easier reading.

**Comment by DEberhardt** [comment]
Sadly, I have no screenshots, the easiest I can differenciate is the line starting with "[-] " should remain Red (as it is the actual error) and the Line starting with "Expecting " _could_ be normal white, no background manipluation intended.

Looking at the linked PR, I am wondering if you wouldn't have already covered that with `FailMessage = 'Gray'`  (if "FailMessage" is the message text starting with "Expecting" 😄

**Comment by fflaten** [comment]
> I added FailMessage color. Unclear where this would be changed to address the messages you are seeing, or which color to pick.

https://github.com/pester/Pester/blob/3c1b60138c1cde3a51c8db333de41a3920422fe1/src/functions/Output.ps1#L738 and 
https://github.com/pester/Pester/blob/3c1b60138c1cde3a51c8db333de41a3920422fe1/src/functions/Output.ps1#L607-L608

> Sadly, I have no screenshots, the easiest I can differenciate is the line starting with "[-] " should remain Red (as it is the actual error) and the Line starting with "Expecting " could be normal white, no background manipluation intended.

Something like this? (using: `& $SafeCommands['Write-Host'] -ForegroundColor 'Gray' "$withMargin"`)

![image](https://user-images.githubusercontent.com/3436158/97786808-5f2d5d00-1bae-11eb-892b-c51a9bb195e1.png)

**Comment by DEberhardt** [comment]
Yes! - White is good. Denotes default Output and easy to read 😃  

In the immortal words of the most famous Kazakh who ever didn't live: "Very nice!"

**Comment by fflaten** [comment]
Would've liked it a bit more gray to avoid it blending too much with the summary when there's a lot text like the example below, but unfortunately we're limited to the small selection of ConsoleColors.
![image](https://user-images.githubusercontent.com/3436158/97790366-7a0ccb00-1bc8-11eb-935f-1d7fed41ebe1.png)

It does look clean, but some people might prefer the red. Could consider making the colors configurable like @nohwnd mentioned. Some type of dictionary-option to override specific themecolors.

**Comment by asears** [comment]
Would be good idea to make it customizable.  For new feature, PesterConfiguration.ReportTheme object could be introduced and promoted from $script:ReportTheme.  $script:ReportStrings could be promoted also.  Might be a better way to do this?
Output.ps1 - $script:ReportStrings could be added to remove any more hard-coded strings in Output.ps1.

There are some output experiences which may not leverage this output style (VSCode Extensions, Azure Devops Test Results Publish Task).  

White blends the failmessage with the completed message.  Gray seems better since it doesn't blend with the next message.  Red blends with the error title and suggestion was to remove the sea of red.  Gray wins for now?  Vote on White or Gray?

Could merge this PR if it looks ok and can open another PR for the custom Report Theme feature / Dark theme option Pester?

**Comment by nohwnd** [comment]
I don't get the motivation behind this. The white / gray suggests that this is not part of the error, while it's the part that has the most important information: Why the test failed. And where the test failed.

I can imagine that we make the error message red, up to the first line of stacktrace (first at). And make the rest of the stack trace dark red. That way the still might be perceived as a single message, but making the error white I don't agree with. 

Adding the colors to config I would also like to avoid for now. I am not yet sure what to do with plugins and how to present their configuration. If anything the config should go into the Output section on the configuration object, and possibly Theme sub-section. Right now there is only Verbosity.

**Comment by fflaten** [comment]
> Adding the colors to config I would also like to avoid for now. I am not yet sure what to do with plugins and how to present their configuration. If anything the config should go into the Output section on the configuration object, and possibly Theme sub-section. Right now there is only Verbosity.

I agree. Since output is a plugin, it would be smart to wait until a public plugin-API (incl. configuration-support) is available. However if we consider Output a critical/non-optional plugin even in the future, then  `Output.ReportTheme` isn't ruled out yet. So I guess that's the big question, combined with the urgency to decide now, hehe 🙂

**Comment by nohwnd** [comment]
I think it will be something along those lines, some plugins are simply better than others. Code Coverage is also a plugin, and it has a first class config object on the configuration. It's just that the current options for output that are proposed here are super specific to the way we write output now, while the output.verbosity is purposfully vague about the output to allow different interpretation of the settings by different plugins (if someone decides to write one).

**Comment by DEberhardt** [comment]
Not urgent, I would think. Distinguishing the Main line from the explanation could be done in multiple ways. I personally would be happy with Dark Red/Red combination as well as Red/White or Red/Gray, but yes, if this is handled via a plugin, then it should be addressed there.
Shelve it if you want (don't know how to do this with issues on github though^^)

**Comment by nohwnd** [comment]
We parked this pending a public plugin API and a proper Output theme section on the configuration — output is itself a plugin, so bolting on ad-hoc color options wasn't the right move. You noted at the time it wasn't urgent and were happy to shelve it. Five years on with no further pull, I'm closing it; theming can come back as part of the plugin/theme work if there's demand.
